lib/puppet/pops/types/types.rb

```diff
@@ -603,6 +671,111 @@ def _assignable?(o)
       end
     end
 
+    class PIterableType < PAnyType
+      attr_reader :element_type
+
+      def initialize(type)
+        @element_type = type
+      end
+
+      def instance?(o)
+        if @element_type.nil? || @element_type.assignable?(PAnyType::DEFAULT)
+          # Any element_type will do
+          case o
+          when Iterable, String, Hash, Enumerable, Enumerator, Range, PEnumType
+            true
+          when Integer
+            o > 0
+          when PIntegerType
+            o.enumerable?
+          else
+            false
+          end
+        else
+          assignable?(TypeCalculator.infer(o))
+        end
+      end
+
+      def generalize
+        @element_type.nil? ? DEFAULT : PIterableType.new(@element_type.generalize)
+      end
+
+      def hash
+        67 * @element_type.hash
+      end
+
+      def iterable?
+        true
+      end
+
+      def iterable_type
+        self
+      end
+
+      def ==(o)
+        self.class == o.class && @element_type == o.element_type
+      end
+
+      DEFAULT = PIterableType.new(nil)
+
+      protected
+
+      # @api private
+      def _assignable?(o)
+        if @element_type.nil? || @element_type.assignable?(PAnyType::DEFAULT)
+          # Don't request the iterable_type since it might be expensive to compute the
+          # type of its elements
+          o.iterable?
+        else
+          o = o.iterable_type
+          o.nil? || o.element_type.nil? ? false : @element_type.assignable?(o.element_type)
+        end
+      end
+    end
+
+    # @api public
+    #
+    class PIteratorType < PAnyType
+      attr_reader :element_type
+
+      def initialize(type)
+        @element_type = type
+      end
+
+      def instance?(o)
+        o.is_a?(Iterable) && (@element_type.nil? || @element_type.assignable?(o.element_type))
+      end
+
+      def generalize
+        @element_type.nil? ? DEFAULT : PIteratorType.new(@element_type.generalize)
+      end
+
+      def hash
+        71 * @element_type.hash
+      end
+
+      def iterable?
+        true
+      end
+
+      def iterable_type
+        element_type.nil? ? PIteratbleType::DEFAULT : PIterableType.new(@element_type)
+      end
+
+      def ==(o)
+        self.class == o.class && @element_type == o.element_type
+      end
+
+      DEFAULT = PIteratorType.new(nil)
+
+      protected
+
+      # @api private
+      def _assignable?(o)
+        o.is_a?(PIteratorType) && (@element_type.nil? || @element_type.assignable?(o.element_type))
```

Comment:
Because `o.is_a?(PIteratorType)` and `PIteratorType` defines the method `element_type`